File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/pose.py
import math
from typing import Tuple, Optional
from PyQt5.QtCore import pyqtSignal, pyqtSlot
from PyQt5.QtWidgets import QHBoxLayout

from .base import ParamGetterWidget
from .utils import FloatGetter


class ParamGetterWidget_Pose(ParamGetterWidget):
    value_changed = pyqtSignal(float, float, float, float, float, float)

    def __init__(
        self,
        param_name: str,
        description_text: Optional[str] = None,
        decimals: int = 3,
        xyz_min: Tuple[float, float, float] = (-1e9,) * 3,
        xyz_max: Tuple[float, float, float] = (+1e9,) * 3,
        xyz_default: Tuple[float, float, float] = (0.0,) * 3,
        xyz_suffix: str = " m",
        rpy_min: Tuple[float, float, float] = (-math.pi,) * 3,
        rpy_max: Tuple[float, float, float] = (+math.pi,) * 3,
        rpy_default: Tuple[float, float, float] = (0.0,) * 3,
        rpy_suffix: str = " rad",
    ) -> None:
        super().__init__(param_name, description_text)

        # XYZ
        cols_xyz = QHBoxLayout()
        self._rows.addLayout(cols_xyz)

        self._x = FloatGetter("x", decimals, xyz_min[0], xyz_max[0], 0.1, xyz_default[0], xyz_suffix)
        cols_xyz.addWidget(self._x)

        self._y = FloatGetter("y", decimals, xyz_min[1], xyz_max[1], 0.1, xyz_default[1], xyz_suffix)
        cols_xyz.addWidget(self._y)

        self._z = FloatGetter("z", decimals, xyz_min[2], xyz_max[2], 0.1, xyz_default[2], xyz_suffix)
        cols_xyz.addWidget(self._z)

        # RPY
        cols_rpy = QHBoxLayout()
        self._rows.addLayout(cols_rpy)

        self._roll = FloatGetter("roll", decimals, rpy_min[0], rpy_max[0], 0.1, rpy_default[0], rpy_suffix)
        cols_rpy.addWidget(self._roll)

        self._pitch = FloatGetter("pitch", decimals, rpy_min[1], rpy_max[1], 0.1, rpy_default[1], rpy_suffix)
        cols_rpy.addWidget(self._pitch)

        self._yaw = FloatGetter("yaw", decimals, rpy_min[2], rpy_max[2], 0.1, rpy_default[2], rpy_suffix)
        cols_rpy.addWidget(self._yaw)

        self._x.value_changed.connect(self._on_value_changed)
        self._y.value_changed.connect(self._on_value_changed)
        self._z.value_changed.connect(self._on_value_changed)
        self._roll.value_changed.connect(self._on_value_changed)
        self._pitch.value_changed.connect(self._on_value_changed)
        self._yaw.value_changed.connect(self._on_value_changed)

    def x(self) -> float:
        return self._x.get()

    def y(self) -> float:
        return self._y.get()

    def z(self) -> float:
        return self._z.get()

    def roll(self) -> float:
        return self._roll.get()

    def pitch(self) -> float:
        return self._pitch.get()

    def yaw(self) -> float:
        return self._yaw.get()

    def get(self) -> Tuple[float, float, float, float, float, float]:
        return self.x(), self.y(), self.z(), self.roll(), self.pitch(), self.yaw()

    def set(self, src: Tuple[float, float, float, float, float, float]) -> None:
        self._x.set(src[0])
        self._y.set(src[1])
        self._z.set(src[2])
        self._roll.set(src[3])
        self._pitch.set(src[4])
        self._yaw.set(src[5])

    @pyqtSlot(float)
    def _on_value_changed(self, value: float) -> None:
        self.value_changed.emit(self.x(), self.y(), self.z(), self.roll(), self.pitch(), self.yaw())

File: tobas_gui/tobas_setup_assistant/src/param_getters/pose.cpp
#include "tobas_setup_assistant/param_getters/pose.hpp"

namespace gui
{
namespace sa
{
ParamGetterWidget_Pose::ParamGetterWidget_Pose(const QString& param_name, const QString& description_text)
  : super(param_name, description_text)
{
  // XYZ
  const auto cols_xyz = new QHBoxLayout();
  rows_->addLayout(cols_xyz);

  x_ = new DoubleGetter("x");
  x_->setDecimals(kDecimals);
  x_->setSingleStep(kSingleStep);
  x_->setValue(kDefaultValue);
  x_->setSuffix(kXYZSuffix);
  connect(x_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  cols_xyz->addWidget(x_);

  y_ = new DoubleGetter("y");
  y_->setDecimals(kDecimals);
  y_->setSingleStep(kSingleStep);
  y_->setValue(kDefaultValue);
  y_->setSuffix(kXYZSuffix);
  connect(y_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  cols_xyz->addWidget(y_);

  z_ = new DoubleGetter("z");
  z_->setDecimals(kDecimals);
  z_->setSingleStep(kSingleStep);
  z_->setValue(kDefaultValue);
  z_->setSuffix(kXYZSuffix);
  connect(z_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  cols_xyz->addWidget(z_);

  // RPY
  const auto cols_rpy = new QHBoxLayout();
  rows_->addLayout(cols_rpy);

  roll_ = new DoubleGetter("roll");
  roll_->setDecimals(kDecimals);
  roll_->setSingleStep(kSingleStep);
  roll_->setValue(kDefaultValue);
  roll_->setSuffix(kRPYSuffix);
  connect(roll_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  cols_rpy->addWidget(roll_);

  pitch_ = new DoubleGetter("pitch");
  pitch_->setDecimals(kDecimals);
  pitch_->setSingleStep(kSingleStep);
  pitch_->setValue(kDefaultValue);
  pitch_->setSuffix(kRPYSuffix);
  connect(pitch_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  cols_rpy->addWidget(pitch_);

  yaw_ = new DoubleGetter("yaw");
  yaw_->setDecimals(kDecimals);
  yaw_->setSingleStep(kSingleStep);
  yaw_->setValue(kDefaultValue);
  yaw_->setSuffix(kRPYSuffix);
  connect(yaw_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  cols_rpy->addWidget(yaw_);
}

std::pair<Eigen::Vector3d, Eigen::Vector3d> ParamGetterWidget_Pose::getValue() const
{
  return { { x(), y(), z() }, { roll(), pitch(), yaw() } };
}

bool ParamGetterWidget_Pose::setValue(const std::pair<Eigen::Vector3d, Eigen::Vector3d>& src)
{
  const auto& xyz = src.first;
  const auto& rpy = src.second;

  x_->setValue(xyz.x());
  y_->setValue(xyz.y());
  z_->setValue(xyz.z());
  roll_->setValue(rpy.x());
  pitch_->setValue(rpy.y());
  yaw_->setValue(rpy.z());

  return true;
}

double ParamGetterWidget_Pose::x() const
{
  return x_->getValue();
}

double ParamGetterWidget_Pose::y() const
{
  return y_->getValue();
}

double ParamGetterWidget_Pose::z() const
{
  return z_->getValue();
}

double ParamGetterWidget_Pose::roll() const
{
  return roll_->getValue();
}

double ParamGetterWidget_Pose::pitch() const
{
  return pitch_->getValue();
}

double ParamGetterWidget_Pose::yaw() const
{
  return yaw_->getValue();
}

void ParamGetterWidget_Pose::onValueChanged(double)
{
  Q_EMIT valueChanged({ { x(), y(), z() }, { roll(), pitch(), yaw() } });
}
}  // namespace sa
}  // namespace gui
